CASSANDRA-16619: Add originating host ID to sstable metadata (trunk)

jacek-lewandowski opened the pull request.

blambov reviewed (approved):
LGTM.

Please create a new ticket to generalize this mechanism to handle other pieces of metadata that are only meaningful for the originating node.

Commit:
CASSANDRA-16619: Add originating host ID to sstable metadata

- md, me, nb sstables
- Update version to nb, update MetadataSerializerTest
- add missing test cases in MetadataSerializerTest

Co-authored-by: Jakub Zytka <jakub.zytka@datastax.com>
Co-authored-by: Jacek Lewandowski <jacek.lewandowski@datastax.com>

jakubzytka commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1101,28 +1101,40 @@ public static void setIndexRemoved(String keyspaceName, String indexName)
 
         // Look up the Host UUID (return it if found)
-        if (!result.isEmpty() && result.one().has("host_id"))
+        if (result != null && !result.isEmpty() && result.one().has("host_id"))
```
why is `result != null` necessary now?

jacek-lewandowski replied:
just to make static analysis happy

jakubzytka commented on src/java/org/apache/cassandra/db/commitlog/CommitLogReplayer.java:
```diff
@@ -285,11 +286,25 @@ public void runMayThrow()
+        }
+
+        if (!skippedSSTables.isEmpty()) {
+            logger.warn("Origin of {} sstables is unknown or doesn't match the local node; commitLogIntervals for them were ignored", skippedSSTables.size());
```
We can get a big number here after upgrade. Perhaps it makes sense to add some info about that - that this is nothing to worry about?

jacek-lewandowski replied:
hmm... I think I can add more information

jakubzytka commented on src/java/org/apache/cassandra/io/sstable/format/big/BigFormat.java:
```diff
@@ -151,6 +154,7 @@ public SSTableReader open(SSTableReaderBuilder builder)
             hasCommitLogLowerBound = version.compareTo("mb") >= 0;
             hasCommitLogIntervals = version.compareTo("mc") >= 0;
             hasAccurateMinMax = version.compareTo("md") >= 0;
+            hasOriginatingHostId = version.matches("(m[e-z])|(n[b-z])");
```
`o*` versions are good too ;)
Noone's gonna remember to change this if `o` is ever introduced/
I think we need to make some comparisons with `version.compareTo("nb") >= 0` being the last

jakubzytka commented on test/distributed/org/apache/cassandra/distributed/test/metrics/StreamingMetricsTest.java:
```diff
@@ -140,15 +144,20 @@ public void testMetricsWithRepairAndStreamingToTwoNodes() throws Exception
         testMetricsWithStreamingToTwoNodes(true);
     }
 
+    private int getNumberOfSSTables(Cluster cluster, int node) {
```
could you please elaborate what do these changes fix?

jacek-lewandowski replied:
hehe... this is in-jvm distributed test. I assumed exact number of sstables in various stages of the test case. After adding that sentinel to commit log replayer, there is one extra sstable created because local host id is unknown. And it is unknown because the server is running in a kind of degraded mode (for speeding up tests, etc).

jakubzytka commented on test/unit/org/apache/cassandra/io/sstable/LegacySSTableTest.java:
```diff
@@ -612,7 +612,7 @@ private static void loadLegacyTable(String tablePattern, String legacyVersion) t
      * </p>
      */
-    @Ignore
+//    @Ignore
```
I believe we want this `@Ignore` back, once the serialization samples have been generated

jacek-lewandowski replied:
ah yes, indeed

jakubzytka commented on test/unit/org/apache/cassandra/io/sstable/metadata/MetadataSerializerTest.java:
```diff
@@ -122,53 +124,46 @@ public File serialize(Map<MetadataType, MetadataComponent> metadata, MetadataSer
-
-    @Test
-    public void testMaReadMc() throws IOException
+    private void testVersions(String... versions) throws Throwable
```
I propose `testOldReadsNew` name (and similarly for `testOldReadsNewMVersions` and `testOldReadsNewNVersions`). This is the crux of the test, but with the current name it's hidden in the very details of the implementation.

jacek-lewandowski replied:
ok

Commit:
apply review comments

blerer commented:
Merged manually
